installinator/src/write.rs

```diff
@@ -213,92 +192,69 @@ impl<'a> ArtifactWriter<'a> {
+    ) -> WriteOutput {
         let mut transport = FileTransport;
-        self.write_with_transport(log, &mut transport, event_sender).await
+        self.write_with_transport(cx, log, &mut transport).await
     }
 
     async fn write_with_transport(
         &mut self,
+        cx: &StepContext,
         log: &Logger,
         transport: &mut impl WriteTransport,
-        event_sender: &mpsc::Sender<ReportEvent>,
-    ) -> Vec<M2Slot> {
-        let mut done_drives = Vec::new();
+    ) -> WriteOutput {
+        let mut done_drives = BTreeSet::new();
 
-        loop {
-            // How many drives did we finish writing this iteration?
-            let mut success_this_iter = 0;
+        // How many drives did we finish writing this iteration?
+        let mut success_this_iter = 0;
 
-            // How many drives did we finish writing on a previous iteration?
-            let mut success_prev_iter = 0;
+        // How many drives did we finish writing on a previous iteration?
+        let mut success_prev_iter;
 
-            for (drive, (destinations, progress)) in self.drives.iter_mut() {
-                *progress = match progress {
-                    DriveWriteProgress::Unstarted => {
-                        let new_progress = self
-                            .artifacts
-                            .write_starting_with_host_phase_2(
-                                log,
-                                1,
-                                destinations,
-                                transport,
-                                event_sender,
-                            )
-                            .await;
+        loop {
+            success_prev_iter = success_this_iter;
+            success_this_iter = 0;
 
-                        if new_progress == DriveWriteProgress::Done {
-                            done_drives.push(*drive);
-                            success_this_iter += 1;
-                        }
+            for (drive, (destinations, progress)) in self.drives.iter_mut() {
+                // Register a separate nested engine for each drive, since we
+                // want each drive to track success and failure independently.
+                let write_cx = SlotWriteContext {
+                    log: log.clone(),
+                    artifacts: self.artifacts,
+                    slot: *drive,
+                    destinations,
+                    progress: *progress,
+                };
+                let res = cx
+                    .with_nested_engine(|engine| {
+                        write_cx.register_steps(engine, transport);
+                        Ok(())
+                    })
+                    .await;
 
-                        new_progress
+                match res {
+                    Ok(_) => {
+                        // This drive succeeded in this iteration.
+                        *progress = DriveWriteProgress::Done;
+                        done_drives.insert(*drive);
```

Comment:
Can this push the same drive onto `done_drives` twice? The sequence I'm worried about:

1. First iteration
  a. Writing drive A succeeds; push A on to `done_drives`
  b. Writing drive B fails
  c. End of iteration; `success_this_iter` is 1, so we loop
2. Second iteration
  a. Drive A's progress is `DriveWriteProgress::Done`, so `write_cx.register_steps()` doesn't register any steps - I think this means the engine result will be `Ok(_)`, right? Then we hit this branch again, and push A into `done_drives` a second time (and also increment `success_this_iter`, which doesn't seem right).
  b. Result on drive B is irrelevant
  c. `success_prev_iter` is 1, so we break out of the loop

Maybe we should check for `DriveWriteProgress::Done` at the top of the for loop and `continue` if the drive is done?

Reply:
Hmm, so `done_drives` is a `BTreeSet` and not a `Vec` which means that inserting into it is idempotent.

`success_this_iter` just means that this drive was successful, whether in a previous iteration or in this one.

I added some comments explaining this.

Reply:
Huh, when I wrote my comment I was definitely thinking `done_drives` was a `Vec`. My bad!